src/main/java/org/opentripplanner/transit/service/StopModelBuilder.java

```diff
@@ -1,38 +1,73 @@
 import org.opentripplanner.transit.model.framework.EntityById;
+import org.opentripplanner.transit.model.framework.FeedScopedId;
+import org.opentripplanner.transit.model.framework.ImmutableEntityById;
 import org.opentripplanner.transit.model.site.AreaStop;
+import org.opentripplanner.transit.model.site.AreaStopBuilder;
 import org.opentripplanner.transit.model.site.GroupOfStations;
 import org.opentripplanner.transit.model.site.GroupStop;
+import org.opentripplanner.transit.model.site.GroupStopBuilder;
 import org.opentripplanner.transit.model.site.MultiModalStation;
 import org.opentripplanner.transit.model.site.RegularStop;
+import org.opentripplanner.transit.model.site.RegularStopBuilder;
 import org.opentripplanner.transit.model.site.Station;
 
 public class StopModelBuilder {
 
-  private final EntityById<RegularStop> regularStopById = new EntityById<>();
-  private final EntityById<AreaStop> areaStopById = new EntityById<>();
-  private final EntityById<GroupStop> groupStopById = new EntityById<>();
-  private final EntityById<Station> stationById = new EntityById<>();
-  private final EntityById<MultiModalStation> multiModalStationById = new EntityById<>();
-  private final EntityById<GroupOfStations> groupOfStationById = new EntityById<>();
+  private final StopModel original;
+  private final AtomicInteger indexCounter;
 
-  StopModelBuilder() {}
+  private final EntityById<RegularStop> regularStopById = new DefaultEntityById<>();
+  private final EntityById<AreaStop> areaStopById = new DefaultEntityById<>();
+  private final EntityById<GroupStop> groupStopById = new DefaultEntityById<>();
+  private final EntityById<Station> stationById = new DefaultEntityById<>();
+  private final EntityById<MultiModalStation> multiModalStationById = new DefaultEntityById<>();
+  private final EntityById<GroupOfStations> groupOfStationById = new DefaultEntityById<>();
 
-  StopModelBuilder(StopModel stopModel) {
-    addAll(stopModel);
+  StopModelBuilder(StopModel original) {
+    this.original = original;
+    this.indexCounter = new AtomicInteger(original.stopIndexSize());
   }
 
-  public EntityById<RegularStop> regularStopsById() {
+  public StopModel original() {
+    return original;
+  }
+
+  int stopIndexSize() {
+    return indexCounter.get();
+  }
+
+  public ImmutableEntityById<RegularStop> regularStopsById() {
     return regularStopById;
   }
 
+  public RegularStopBuilder regularStop(FeedScopedId id) {
+    return RegularStop.of(id, indexCounter::getAndIncrement);
+  }
+
+  public RegularStop computeRegularStopIfAbsent(
+    FeedScopedId id,
+    Function<FeedScopedId, RegularStop> factory
+  ) {
+    return regularStopById.computeIfAbsent(id, factory);
+  }
+
   public StopModelBuilder withRegularStop(RegularStop stop) {
     regularStopById.add(stop);
     return this;
   }
 
-  public EntityById<Station> stationById() {
+  public StopModelBuilder withRegularStops(Collection<RegularStop> stops) {
```

Comment:
As dicussed with @t2gran these `withXX` methods won't be necessary in the future.